src/redismodule.h

```diff
@@ -34,6 +34,12 @@ typedef long long ustime_t;
 #define REDISMODULE_OK 0
 #define REDISMODULE_ERR 1
 
+/* Custom Auth status return values. */
+#define REDISMODULE_AUTH_SUCCEEDED 0
+#define REDISMODULE_AUTH_DENIED 1
```

Comment:
Repost: I also don't know if we need both of these return codes, I might combine them into one. Whether or not auth succeeded should depend on whether or not the user was authenticated. Thinking more, maybe we should combine SUCCEEDED, DENIED, and BLOCKED together. The behavior being moduled by the APIs that were called.

Reply:
If we reduce the 4 codes into 2 codes, we can have: `REDISMODULE_AUTH_HANDLED` and `REDISMODULE_AUTH_NOT_HANDLED`.

To authenticate, the module can use the `RM_Authenticate*` APIs and return `REDISMODULE_AUTH_HANDLED`.
To block a client on module based Auth, the module can use the `RedisModule_BlockClientOnAuth` API and return `REDISMODULE_AUTH_HANDLED`.
To deny authentication, the module will need to return `REDISMODULE_AUTH_HANDLED` without successfully using either of the APIs above.
To indicate that the Module does not want to take any action related to AUTH, the module can return `REDISMODULE_AUTH_NOT_HANDLED`. 


This works fine - although denying authentication from the module is not very explicit since we don't have a specific return code or an API to deny auth.

I can make the update to reduce the 4 codes to 2 (`REDISMODULE_AUTH_HANDLED` and `REDISMODULE_AUTH_NOT_HANDLED`) if this sounds good

Reply:
> This works fine - although denying authentication from the module is not very explicit since we don't have a specific return code or an API to deny auth.

That's true. They can be explicit by calling the DenyAuth api if they want, but if they call handled but don't do anything. I think this behavior, if you aren't authenticated you then throw an error is OK. However, one nuance is if you are using `HELLO` or `AUTH` on an *already* authenticated user, we won't be able to detect anything. Perhaps we do need REDISMODULE_AUTH_DENIED then.

Reply:
> However, one nuance is if you are using HELLO or AUTH on an already authenticated user, we won't be able to detect anything.

We are able to differentiate using the client flag. The `CLIENT_CUSTOM_AUTH_RESULT` will be momentarily added to the client flags when the module uses the `RM_Auth*` APIs & is removed right before replying to the client.

If this flag is missing, we will know that the Module did not successfully authenticate the client with custom auth and in this case, we can reply to the client with an ERR message if the module returned `REDISMODULE_AUTH_HANDLED`

Reply:
I have switched to using 2 codes in this commit: https://github.com/redis/redis/pull/11659/commits/b21ad243127fddee6889652e579a2d15fa725aed